aion.sh

```diff
@@ -37,5 +37,158 @@ ARG=$@
+      set --
+      first=false
+    else		  
+      set -- "$@" "$arg"
+    fi
+  done
+  guard=true
+fi
+
+if $guard; then
+
+	echo "######################"
+	echo "  Watchguard Enabled  "
+	echo "######################"
+
+	wait=300	# sec
+	sample=1	# sec
+	tolerance=1 	# sec
+	threadRate=1 	# rate
+
+	config=config/config.xml
+	logging=$(egrep -o "log-file.*log-file" $config | cut -d">" -f2 | cut -d"<" -f1)
+	logpath=$(egrep -o "log-path.*log-path" $config | cut -d">" -f2 | cut -d"<" -f1)
+	file=$logpath/aionCurrentLog.dat
+
+	noInterrupt=true
+	countRebounce=0
+	running=false
+	watching=false
+	lastBoot=0
+
+	trap "exit" INT TERM
+	trap "interrupt" EXIT
+	function interrupt() {
+
+		# Interrupts the Aion kernel and awaits shutdown complete
+		if $running; then
+		  kill $kPID
+		  temp=$(ps --pid $kPID | egrep -o "$kPID")
+		  while [[ $temp -eq $kPID ]] ; do
+		    sleep 2s
+		    temp=$(ps --pid $kPID | egrep -o "$kPID")
+		  done
+		  running=false
+		  watching=false
+		  noInterrupt=false
+		fi
+
+		# Interrupts the watchguard (current process)
+		kill $$
+
+	}
+
+	# Keep executing aion kernel until interrupted
+	while $noInterrupt; do
+
+		# Reboot timer
+		newBoot=$(date +"%s")
+		let "duration=$newBoot-$lastBoot"
+		if [[ $duration -lt $wait ]]; then
+		  let "sleep=$wait-$duration"
+		  echo "[Reboot Timer: $sleep]"
+		  sleep $sleep
+		fi
+		lastBoot=$newBoot
+
+		# Execute Java kernel
+		env EVMJIT="-cache=1" ./rt/bin/java -Xms4g \
+			-cp "./lib/*:./lib/libminiupnp/*:./mod/*" org.aion.Aion "$@" &
+		kPID=$!
+		running=true
+		watching=true
+		checkRate=0
+		tPrev=0
+
+		# Watchguard
+		while $watching; do
+
+			sleep $sample
+
+			# [1] Log timestamp (last 60 sec) OR [2] PID process state ZOMBIE/DEAD
+			if $logging; then
+			  last=$(stat $file | egrep "Modify\:" | cut -d" " -f3 | cut -d"." -f1)
+			  lastUTC=$(date --date="$last" +"%s")
+			  nowUTC=$(date +"%s")
+			  if [[ $lastUTC -lt $((nowUTC-$tolerance)) ]] || (ps $kPID | cut -c 16-22 | egrep -v "STAT" | egrep -q "Z"); then
+			    echo "## KERNEL DEAD FOR $tolerance SEC ##"
+			    watching=false
+			  fi
+			fi
+
+			# [1] Thread runtime (last 60 sec) AND [2] PID thread state BLOCKED
+			if [ $checkRate -eq $threadRate ]; then
+			  let "duration=$sample*$threadRate"
+			  temp='p2p-in sync-ib'
+			  threads=($temp)
+			  checkRate=0
+			  for ((i=0; i<${#threads[@]}; ++i)); do
+			    tTime=$(ps --pid $kPID | egrep -o "[0-9]{2}\.[0-9]{2} ${threads[i]}" | cut -d" " -f1)
+			    tState=$(jstack -l $kPID | egrep -A1 "${threads[i]}" | egrep -o "State.*" | cut -d" " -f2)
+			    if [[ $tTime == ${tPrev[i]} ]] && [[ $tState == "BLOCKED" ]]; then 
+			      echo "## ${threads[i]} THREAD DEAD ##"
+			      (jstack -l $kPID | egrep -A1 "${threads[i]}") > threadDump_$countRebounce.txt
+			      watching=false
+			    fi
+			    tPrev[i]=$tTime
+			  done
+			else
+			  ((checkRate++))
+			fi
+
+		done
+	
+		# Shutsdown Aion kernel
+		echo "## Killing Kernel ##"
+		kill $kPID
```

Comment:
Question: is it possible for this script to get stuck if the kernel does _not_ respond to a kill (``SIGINT``?) request.

Is there anywhere that was escalate to a ``SIGKILL`` (for example if the kernel is still alive after 1 minute)

Reply:
ill add a shutdown timer to force the script to kill the process